lib/Scheduling/CPSATSchedulers.cpp

```diff
@@ -0,0 +1,154 @@
+
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/Format.h"
+
+#define DEBUG_TYPE "cpsat-schedulers"
+
+using namespace circt;
+using namespace circt::scheduling;
+using namespace operations_research;
+using namespace operations_research::sat;
+
+using llvm::dbgs;
+using llvm::format;
+
+/// Solve the shared operators problem by modeling it as a Resource
+/// Constrained Project Scheduling Problem (RCPSP), which in turn is formulated
+/// as a Constraint Programming (CP) Satisfiability (SAT) problem.
+///
+/// This is a high-fidelity translation of
+/// https://github.com/google/or-tools/blob/stable/examples/python/rcpsp_sat.py
+/// but a gentler introduction (though with differing formulation) is
+/// https://python-mip.readthedocs.io/en/latest/examples.html
+LogicalResult scheduling::scheduleCPSAT(SharedOperatorsProblem &prob,
+                                        Operation *lastOp) {
+  Operation *containingOp = prob.getContainingOp();
+  if (!prob.hasOperation(lastOp))
+    return containingOp->emitError("problem does not include last operation");
+
+  CpModelBuilder cpModel;
+  auto &tasks = prob.getOperations();
+  auto &resources = prob.getOperatorTypes();
+
+  DenseMap<Operation *, IntVar> taskStarts;
+  DenseMap<Operation *, IntVar> taskEnds;
+  DenseMap<Problem::OperatorType, SmallVector<IntervalVar, 4>>
+      resourcesToTaskIntervals;
+
+  // First get horizon, i.e., the time taken if all operations were executed
+  // sequentially.
+  unsigned horizon = 0;
+  for (auto *task : tasks) {
+    unsigned duration = *prob.getLatency(*prob.getLinkedOperatorType(task));
+    horizon += duration;
+  }
+
+  // Build task-interval decision variables, which effectively serve to
+  // constrain startVar and endVar to be duration apart. Then map them
+  // to the resources (operators) consumed during those intervals. Note,
+  // resources are in fact not constrained to be occupied for the whole of the
+  // task interval, but only during the first "tick". See comment below
+  // regarding cpModel.NewFixedSizeIntervalVar.
+  for (auto item : llvm::enumerate(tasks)) {
+    auto i = item.index();
+    auto *task = item.value();
+    IntVar startVar = cpModel.NewIntVar(Domain(0, horizon))
+                          .WithName((Twine("start_of_task_") + Twine(i)).str());
+    IntVar endVar = cpModel.NewIntVar(Domain(0, horizon))
+                        .WithName((Twine("end_of_task_") + Twine(i)).str());
+    taskStarts[task] = startVar;
+    taskEnds[task] = endVar;
+    auto resource = prob.getLinkedOperatorType(task);
+    unsigned duration = *prob.getLatency(*resource);
+    IntervalVar taskInterval =
+        cpModel.NewIntervalVar(startVar, duration, endVar)
+            .WithName((Twine("task_interval_") + Twine(i)).str());
+
+    if (prob.getLimit(*resource))
+      resourcesToTaskIntervals[resource.value()].emplace_back(taskInterval);
+  }
+
+  // Check for zero-distance cycles and otherwise establish operation ordering
```

Comment:
Drop the "zero-distance" (here and a few lines below) please, that would only make sense in a cyclic problem.